### ISSUE TITLE
WAL files not collected in conjunction with long running transactions

### ISSUE THREAD

**Issue by majuss** [arangodb/arangodb#2635]
### my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [x] 3.1

On this operating system:
Latest Debian with EXT4.

---

This problem occurs everytime when I restart arangodb with: `sudo service arangodb3 restart`.
The database is idling, CPU usage of the computer is ~1%. As soon as I run the above command the arangod uses 100% CPU for 6 Minuten between 17:53 and 17:59 (see log). Needless to say that everything is unresponsive.

This is what the log yields:
```
2017-06-21T17:52:56Z [11705] INFO control-c received, beginning shut down sequence
2017-06-21T17:53:57Z [17010] INFO ArangoDB 3.1.21 [linux] 64bit, using VPack 0.1.30, ICU 54.1, V8 5.0.71.39, OpenSSL 1.0.1t  3 May 2016
2017-06-21T17:53:57Z [17010] INFO file-descriptors (nofiles) hard limit is 131072, soft limit is 131072
2017-06-21T17:53:57Z [17010] WARNING fcntl on lockfile '/var/lib/arangodb3/LOCK' failed: not implemented. a possible reason is that the filesystem does not support file-locking
2017-06-21T17:53:57Z [17010] FATAL database is locked, please check the lock file '/var/lib/arangodb3/LOCK'
2017-06-21T17:59:09Z [11705] INFO ArangoDB has been shut down
```

After that the arangod is just not running and I have to start it again manually: `sudo service arangodb3 start`

**Comment by dothebart** [comment]
Hi, 
This was a bug in versions previous to 3.1.21 - they would loose the connection from init script to running database instance, and thus fail to automatically stop the old arangodb on upgrade. 

You need to stop ArangoDB Manually once:

`killall arangod`

after that the upgrade / `dpkg --pending --configure ` will run through.

**Comment by jsteemann** [comment]
I don't think that explains the 5+ minutes with 100% CPU usage before shutting down.

**Comment by jsteemann** [comment]
From the above log output it looks like there has been an `arangod` process with pid 11705 which received the shut down command at 17:52:56Z:
```
2017-06-21T17:52:56Z [11705] INFO control-c received, beginning shut down sequence
```
This instance (11705) was then performing the shutdown.
Then the log shows the start of a new instance with pid 17010 at 17:53:57Z:
```
017-06-21T17:53:57Z [17010] INFO ArangoDB 3.1.21 [linux] 64bit, using VPack 0.1.30, ICU 54.1, V8 5.0.71.39, OpenSSL 1.0.1t  3 May 2016
```
Starting the instance with pid 17010 failed because the instance with pid 11705 was still running:
```
2017-06-21T17:53:57Z [17010] FATAL database is locked, please check the lock file '/var/lib/arangodb3/LOCK'
```
Finally the initial instance with pid 11705 completed its shutdown at 17:59:09Z:
```
2017-06-21T17:59:09Z [11705] INFO ArangoDB has been shut down
```
In the end, no arangods will be running, as the initial instance was shut down and starting the next instance was unsuccessful.

This may be a problem with the init script. We'll check that.

**Comment by dothebart** [comment]
Ok, first bug is that the init script doesn't abort if it fails to shut down the old arangod: (the ||exit 1 is new)
```
  restart)
    $0 stop || exit 1
    sleep 3
    $0 start
    exit $?
    ;;
```

**Comment by jsteemann** [comment]
Is the long shutdown time reproducible, or did it happen just that one time?

**Comment by majuss** [comment]
It happens everytime and is 100% reproducible. I can upload you the full 100GB VM qcow2 image or all the logs if you want to have a closer look.

**Comment by jsteemann** [comment]
What would be helpful as a first step would be logfiles from the point of time the server receives the shutdown command until it actually shuts down.
Can you issue the following curl command before the shutdown, which increases the log verbosity a lot?

```
curl -X PUT http://127.0.0.1:8529/_admin/log/level --data '{"general":"trace","collector":"trace","compactor":"trace","datafiles":"trace","mmap":"trace","performance":"trace","queries":"trace","requests":"trace","startup":"trace","syscall":"trace","v8":"trace","threads":"trace"}' --dump -- --basic --user "root:"
```
Note that IP address, port, and/or username/password may need to be adjusted for it to work.
After the log verbosity has been changed, can you wait until the actual shutdown of the server and after that provide the server's logfile? It should hopefully provide a lot more details so we can see which phases take long to complete. 
Thanks!

**Comment by majuss** [comment]
Typical story... as soon as I changed the verbosity the db restarted just fine...
Maybe you can still find something in the log.

[arangod.log.zip](https://github.com/arangodb/arangodb/files/1094995/arangod.log.zip)

**Comment by majuss** [comment]
So the problem seems to be back after I restarted my whole VM. Now the arangod got 100% and is doing this:

```
2017-06-22T21:29:56Z [617] INFO replaying WAL logfile '/var/lib/arangodb3/journals/logfile-2576354758.db' (688 of 695)
2017-06-22T21:29:58Z [617] INFO replaying WAL logfile '/var/lib/arangodb3/journals/logfile-2576596089.db' (689 of 695)
2017-06-22T21:30:00Z [617] INFO replaying WAL logfile '/var/lib/arangodb3/journals/logfile-2576837481.db' (690 of 695)
2017-06-22T21:30:02Z [617] INFO replaying WAL logfile '/var/lib/arangodb3/journals/logfile-2577071437.db' (691 of 695)
2017-06-22T21:30:04Z [617] INFO replaying WAL logfile '/var/lib/arangodb3/journals/logfile-2577306469.db' (692 of 695)
2017-06-22T21:30:06Z [617] INFO replaying WAL logfile '/var/lib/arangodb3/journals/logfile-2577549581.db' (693 of 695)
2017-06-22T21:30:08Z [617] INFO replaying WAL logfile '/var/lib/arangodb3/journals/logfile-2577785007.db' (694 of 695)
2017-06-22T21:30:09Z [617] INFO replaying WAL logfile '/var/lib/arangodb3/journals/logfile-2578022657.db' (695 of 695)
2017-06-22T21:30:09Z [617] INFO WAL recovery finished successfully
2017-06-22T21:30:22Z [617] INFO JavaScript using startup '/usr/share/arangodb3/js', application '/var/lib/arangodb3-apps'
```

After he's done I have to start the deamon manually again. Then it's still recovering WAL files but ~20 less. Then when he's done he crashes again... This game repeats 6-7 times then I can run arangodb just normally.

**Comment by jsteemann** [comment]
The fact that it replays 695 WAL logfiles at start is obviously not good, because that will cause a very long startup phase for the server.
Do you have any special workload that would explain that high number of (uncollected) WAL files, i.e. are you doing something that would lead to lots of WAL files being created? Possible reasons for this are:
- WAL files are too small (default is 32 MB) to hold a reasonable number of documents. If you work with huge documents (>= 1MB) then this would be a possible explanation
- are you using very long-running write transactions? This could cause WAL files to pile up. They will be collected by the WAL collector once the transaction commits, but if one transaction is very long-running it would block the WAL collector and the number of WAL files could increase
Could one of the two reasons apply in your case?

Additionally you mentioned that the server crashes after WAL recovery. I would like to know why, because that is obviously unintended. Do you know if the process gets killed (e.g. by the system's OOM killer because it consumes almost all of the available RAM) or really crashes? If it crashes, is there a chance to enable core dumps so we can see where exactly it crashes?
A brief description for retrieving core dumps can be found [https://github.com/arangodb/arangodb/blob/devel/README_maintainers.md#linux-coredumps](here).

Thanks!

**Comment by majuss** [comment]
Yes I was running several transactions that took longer than 45 minutes. 

Where can I find the debian debug symbols for 3.1.21? 
−−-

So I can't use the DB at the moment. Everytime it starts it's recovering WAL files. I let it run the whole night (with a while true loop, service arangodb3 start) but there are still 350 left. Something is completely broken.

**Comment by jsteemann** [comment]
Is it possible to split the transactions that lasted 45+ minutes into multiple smaller ones?
That should fix the problem of too many WAL files piling up.

**Comment by majuss** [comment]
So once I started the arangod manually the recovery of log files was gone... I reviewed my code: I have no long transaction in it, only recursive async functions that are running several minutes.

**Comment by baslr** [comment]
@jsteemann what I dont understand is this: if arangod idles for a while, days or so, it should start compacting the wal into datafiles. I mean ~1k wal files are not compacted since then.

**Comment by jsteemann** [comment]
WAL file compaction will halt when there are still transactions running that could reach data in one of those WAL files.
If transactions don't last long, then compaction shouldn't stall.
@majuss : the "recursive async functions" you mentioned are in the client application or in server-side (Foxx) JavaScript code? Do they use any transactions?

**Comment by majuss** [comment]
It's client-sides javascript code. This: https://github.com/majuss/globi-parasites/blob/master/build_freeliving_source.js for example.

After I run my complete pipeline there are 990 WAL files, arangod is idling since 3+hours.

**Comment by jsteemann** [comment]
A few additional questions: 
* can you, at any point after the server has properly started, run `curl -X GET http://127.0.0.1:8529/_admin/wal/properties --dump - --basic --user "root:passwd"` (with the actual IP/port/username/password) and post the response here? That will show the WAL configuration of the server.
* is the server process showing some I/O or CPU usage in top after you ran your pipeline and there are that many WAL files? Or is it completely idle? That would be interesting to see. It is expected that it will collect WAL files and write data from WAL files to collection datafiles. 
* after you ran your pipeline, can you also run `curl -X GET http://127.0.0.1:8529/_admin/wal/transactions --dump - --basic --user "root:passwd"` (again with the actual IP/port/username/password) and post the response here? That will show whether there are still pending transactions that will block WAL collection.

Thanks!

**Comment by dothebart** [comment]
@majuss any news on this? can you still reproduce this with arangodb 3.2?

**Comment by dothebart** [comment]
Closing as solved with 3.2.8
